src/esphomelib/sensor/tx20_component.cpp

```diff
@@ -0,0 +1,160 @@
+      wind_direction_text_sensor_(new TX20WindDirectionTextSensor(wind_direction_text_name)),
+      pin_(pin) {
+  ESP_LOGV(TAG, "Start TX20Component...");
+  if(firstInstance){
+    ESP_LOGE(TAG, "Multiple TX20 Sensors are not supported!!");
+  }
+  firstInstance = true;
+}
+
+void TX20Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up TX20...");
+  this->pin_->setup();
+
+  attachInterrupt(this->pin_->get_pin(), TX20Component::start_read_, RISING);
+  global_tx20_ = this;
+}
+void TX20Component::dump_config() {
+  ESP_LOGCONFIG(TAG, "TX20:");
+  LOG_PIN(" Pin:", TX20Component::pin_);
+  LOG_UPDATE_INTERVAL(this);
+}
+float TX20Component::get_setup_priority() const {
+  return setup_priority::HARDWARE_LATE;
+}
+
+void TX20Component::update() {
+  if (tx20_available) {
+    ESP_LOGV(TAG, "Updating TX20...");
+    std::string direction = DIRECTIONS.at(tx20_wind_direction);
+
+    this->wind_direction_sensor_->publish_state(tx20_wind_direction);
+    this->wind_speed_sensor_->publish_state(tx20_wind_speed_kmh);
+    this->wind_direction_text_sensor_->publish_state(direction);
+  }
+}
+void TX20Component::start_read_() {
+  global_tx20_->start_read_internal_();
+}
+
+void TX20Component::start_read_internal_() {
+  /* La Crosse TX20 Anemometer datagram every 2 seconds
+   * 0-0 11011 0011 111010101111 0101 1100 000101010000 0-0 - Received pin data
+   * at 1200 uSec per bit sa    sb   sc           sd   se   sf 00100 1100
+   * 000101010000 1010 1100 000101010000     - sa to sd inverted user data,
+   * LSB first sa - Start frame always 00100 sb - Wind direction 0 - 15 sc -
+   * Wind speed 0 - 511 sd - Checksum se - Wind direction 0 - 15 sf - Wind speed
+   * 0 - 511
+   */
+
+  tx20_available = false;
+
+  uint8_t tx20_sa = 0;
+  uint8_t tx20_sb = 0;
+  uint8_t tx20_sd = 0;
+  uint8_t tx20_se = 0;
+  uint8_t tx20_sc = 0;
+  uint8_t tx20_sf = 0;
+
+  delayMicroseconds(TX20_BIT_TIME / 2);
+
+  for (int bitcount = 41; bitcount > 0; bitcount--) {
+    uint8_t dpin = (digitalRead(this->pin_->get_pin()));
+    if (bitcount > 41 - 5) {
+      // start, inverted
+      tx20_sa = (tx20_sa << 1) | (dpin ^ 1);
+    } else if (bitcount > 41 - 5 - 4) {
+      // wind dir, inverted
+      tx20_sb = tx20_sb >> 1 | ((dpin ^ 1) << 3);
+    } else if (bitcount > 41 - 5 - 4 - 12) {
+      // windspeed, inverted
+      tx20_sc = tx20_sc >> 1 | ((dpin ^ 1) << 11);
+    } else if (bitcount > 41 - 5 - 4 - 12 - 4) {
+      // checksum, inverted
+      tx20_sd = tx20_sd >> 1 | ((dpin ^ 1) << 3);
+    } else if (bitcount > 41 - 5 - 4 - 12 - 4 - 4) {
+      // wind dir
+      tx20_se = tx20_se >> 1 | (dpin << 3);
+    } else {
+      // windspeed
+      tx20_sf = tx20_sf >> 1 | (dpin << 11);
+    }
+
+    delayMicroseconds(TX20_BIT_TIME);
+  }
+
+  uint8_t chk = (tx20_sb + (tx20_sc & 0xf) + ((tx20_sc >> 4) & 0xf) + ((tx20_sc >> 8) & 0xf));
+  chk &= 0xf;
+
+  if ((chk == tx20_sd) && (tx20_sc < 400)) {  // if checksum seems to be ok and wind speed below 40 m/s
```

Comment:
Is there a reason wind speeds above 40m/s are discarded?

Reply:
51.1 is the maximum that is supported by the sensor. I have not seen mine go anywhere close to 40 - so my guess is that it is not working so well in that range.

Reply:
Ok so the sensor has a checksum, so that should be enough for checking if everything's alright. Maybe somebody else _has_ conditions where the wind sensor gets above 40 m/s (up to the 51.1 max).

Can we just drop the second check? I think the checksum by itself is just fine.

Reply:
![image](https://user-images.githubusercontent.com/1206407/52190346-e9f8f080-283e-11e9-9ad9-79337b37467c.png)

Was testing this for half a day - seems to be not good enough. Readded the check again.